Sources/ApodiniOpenAPI/Document/OpenAPIComponentsObjectBuilder.swift

```diff
@@ -101,102 +105,26 @@ class OpenAPIComponentsObjectBuilder {
 extension OpenAPIComponentsObjectBuilder {
     /// Builds the schema for a type and returns it together with a suitable title.
     private func buildSchemaWithTitle(for type: Any.Type) throws -> (JSONSchema, String) {
-        let node: Node<ReflectionInfo> = try Self.node(type)
-        let schemaNode = node.contextMap { (node: Node<ReflectionInfo>) -> JSONSchema in
-            let schema = deriveSchemaFromReflectionInfo(node)
-            let schemaName = createSchemaName(for: node)
-            
-            // If there is a reference type that is not yet saved into the
-            // componentsObject, a new schema is created and saved.
+        let rootTypeInformation = try TypeInformation(type: type)
+        
+        rootTypeInformation.allTypes().forEach { typeInformation in
+            let schema: JSONSchema = .from(typeInformation: typeInformation)
+            let schemaName = typeInformation.jsonSchemaName()
             if schema.isReference && !schemaExists(for: schemaName) {
-                var properties: [String: JSONSchema] = [:]
-                for child in node.children {
-                    if let propertyInfo = child.value.propertyInfo {
-                        properties[propertyInfo.name] = deriveSchemaFromReflectionInfo(child)
-                    }
+                let properties = typeInformation.objectProperties.reduce(into: [String: JSONSchema]()) { result, current in
+                    result[current.name] = .from(typeInformation: current.type)
```

Comment:
Would we, in this case, need to recursively resolve `.reference` JSONSchemes here? Or is this dealt through the `allTypes` above anyways?

Reply:
Yes exactly, there is no need to because `allTypes` returns all distinct types contained in `rootTypeInformation`.

Reply:
thanks for the clarification